db/revision_cache_lru.go

```diff
@@ -912,17 +918,32 @@ func (rc *LRURevisionCache) revCacheMemoryBasedEviction(ctx context.Context) {
 
 // performEviction will evict the oldest items in the cache till we are below the memory threshold
 func (rc *LRURevisionCache) performEviction(ctx context.Context) {
+	numItemsRemoved := rc.evictBasedOffMemoryUsage(ctx)
+	rc.cacheNumItems.Add(-numItemsRemoved)
+}
+
+func (rc *LRURevisionCache) evictBasedOffMemoryUsage(ctx context.Context) int64 {
 	var numItemsRemoved, numBytesRemoved int64
-	rc.lock.Lock() // hold rev cache lock to remove items from cache until we're below memory threshold for the shard
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
 	// check if we are over memory capacity after holding rev cache mutex (protect against another goroutine evicting whilst waiting for mutex above)
 	if currMemoryUsage := rc.currMemoryUsage.Value(); currMemoryUsage > rc.memoryCapacity {
 		// find amount of bytes needed to evict till below threshold
 		bytesNeededToRemove := currMemoryUsage - rc.memoryCapacity
 		for bytesNeededToRemove > numBytesRemoved {
 			value := rc._findEvictionValue()
 			if value == nil {
-				// no more values ready for eviction
-				break
+				if rc.lruList.Len() > 0 {
+					// no more values ready for eviction
+					break
+				} else {
+					// list is empty, nothing more to evict but stats are wrong so zero stats and return
+					base.DebugfCtx(ctx, base.KeyCache, "Revision cache memory stats inconsistent for this shard, resetting to zero")
+					correctionVal := rc.currMemoryUsage.Value()
+					rc.currMemoryUsage.Add(-correctionVal)
+					rc.cacheMemoryBytesStat.Add(-correctionVal)
```

Comment:
It would be much clearer and potentially safer to use Set(0) for this than doing two separate Value and Add IMO